### PR TITLE
Release candidate build for vite! API

### PR DESCRIPTION
**Pull request by atyrode**
This pull request encompasses the following features:

- #1
- #2
- #4
- #6
- #8

And some miscellaneous additions as well as a web page for the API were implemented in:

- #10.

All these features makes **vite!** functional, hopefully robust and thereby make the current implementation probably a good release candidate 😄 

All the features have a test suite, and relevant comments were left in the code to aid in maintaining this project.

~~By default, it will shorten the links to the **domain name** `http://vite.lol/` whose links will of course only work in the context of being a subdomain of the host.~~ 
edit: not anymore, it nows depend on two secrets (protocol and host)

### CODE REVIEW

**Comment by atyrode** [comment]
Tested to be working and running by following the readme in a fresh ubuntu docker container, mapping the ports 8080:8080 with the extra step of having to git checkout to dev for now.

File: src/api.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        url = url[len(DOMAIN_NAME):]
+    elif url.startswith(SHORT_URL):
+        url = url[len(SHORT_URL):]
+        
+    if url == "":
+        return {"error": "No URL provided"}
+    elif url == "0": # There's no row id 0, so there can't be a shortened URL for it
+        return {"url": "https://en.wikipedia.org/wiki/0#Computer_science", "clicks": -1}
+    elif url_charset.validate(url) == False:
+        return {"error": "Not a valid URL"}
```
Comment (jbbeuzelin) [inline src/api.py:90]: these checks need to be performed earlier because the previous test uses the url parameter, so if it's null, you'll never get to your test

File: src/api.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        url (str): The URL to decode
+
+    Returns:
+        dict: A JSON response containing the original URL and the number of clicks
+    """
+        
+    if url.startswith(DOMAIN_NAME):
+        url = url[len(DOMAIN_NAME):]
+    elif url.startswith(SHORT_URL):
+        url = url[len(SHORT_URL):]
```
Comment (jbbeuzelin) [inline src/api.py:83]: use `ternary` for readability
```suggestion
clearUrl = url[len(DOMAIN_NAME):] if url.startswith(DOMAIN_NAME) else (url[len(SHORT_URL):] if url.startswith(SHORT_URL): else url);

```

File: src/api.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    original_url = decode_result["url"]
+        
+    with ViteDB(DB_PATH) as db:
+        decoded_id: int = codec.decode(url)
+        db.increment_clicks(decoded_id)
+    
+    if not original_url.startswith("http"):
+        original_url = "http://" + original_url
+        
+    return RedirectResponse(original_url)
```
Comment (jbbeuzelin) [inline src/api.py:146]: strange purpose of this controler

File: src/charset.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    def __str__(self) -> str:
+        return self.charset
+    
+    def validate(self, input: str) -> bool:
+        """Returns a boolean indicating if the input is valid for the charset."""
+        
+        if set(input).difference(self.charset) != set():
+            return False
+        
+        return True
```
Comment (jbbeuzelin) [inline src/charset.py:66]:
```suggestion
        return  set(input).difference(self.charset) == set():
```

File: src/charset.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        if not any(self.__dict__.values()) is True:
+            raise ValueError("At least one charset type must be True for URLCharset")
+    
+    @property
+    def charset(self) -> str:
+        comp: str = str()
+        comp += string.digits           if self.numeric         == True else ''
+        comp += string.ascii_lowercase  if self.lowercase_ascii == True else ''
+        comp += string.ascii_uppercase  if self.uppercase_ascii == True else ''
+        comp += "~_-."                  if self.special         == True else ''
```
Comment (jbbeuzelin) [inline src/charset.py:42]: `== True ` is not Mandatory here ?

File: src/database.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
+import sqlite3
+from dataclasses import dataclass, field
+from typing import Optional, Type, Union, Tuple, List
+from types import TracebackType
+
+class DbManager:
```
Comment (jbbeuzelin) [inline src/database.py:6]: Use an ORM instead of coding your request. It's safer in order to prevent Sql Injection. 
Alchemy is a good start

File: src/static/index.css [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    padding: 2rem;
+    border: double 3px transparent;
+    border-radius: 6px;
+    background-image: linear-gradient(rgb(95, 95, 95), rgb(95, 95, 95)), 
+                      linear-gradient(to right, rgb(123, 149, 233), rgb(214, 98, 179));
+    background-origin: border-box;
+    background-clip: padding-box, border-box;
+    font-size: 24px;
+    outline: none;
+    font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
```
Comment (jbbeuzelin) [inline src/static/index.css:39]: declare Fon-family on top of your html instead of define it evey where.

File: src/charset.py (hunk 5 above)
Comment (atyrode) [reply src/charset.py:42]: `== True` is indeed not mandatory and implicit in python ternaries but I advocate for clearer code intent over implicitness. I agree this is a bit anti-pythonic but I think it's best practice and welcomes contributer from a broader language base.

File: src/api.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    if url == "":
+        return {"error": "No URL provided"}
+    
+    with ViteDB(DB_PATH) as db:
+        row_count = db.get_row_count()
+        
+    unique_id: int = row_count + 1
+    encoded_uid: str = codec.encode(unique_id)
+    
+    shortened_url: str = f"{DOMAIN_NAME}{encoded_uid}"
```
[the remark this answers is not in the document]
Comment (atyrode) [reply src/api.py]: This is intended by design, there was an attempt at obsfuscation in a now stale branch, but after inspecting other URL shortener, I observed they were indeed starting at `/0` (e.g: [t.co/0](t.co/0)) so they could better exploit the ability to indeed: shorten.

I've decided to give up on obfuscation because this would entail one of the two following choices:
- longer shortened (ah!) links
- 916 millions possibility on 5 characters obfuscated, this could last a while, but eventually mean the obfuscation would stop making sense after 916 millions generated link.

I've decided traversability of the URL was in fact not an issue if it was a standard. 
I could re-implement the obfuscator but this would be at the cost of longer URLs.

File: src/api.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+from .codec import Codec
+from .database import ViteDB
+
+## CONSTANTS ##
+
+PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+STATIC_PATH  = os.path.join(PROJECT_ROOT, 'src', 'static')
+DB_PATH      = os.path.join(PROJECT_ROOT, 'data', 'vite.db')
+
+DOMAIN_NAME  = "http://vite.lol/"
```
[the remark this answers is not in the document]
Comment (atyrode) [reply src/api.py]: https://github.com/atyrode/vite/pull/12/commits/499bebb26a581ec3e1a80184db1c5c5d681f0310 fixes the https part, regarding the secret management and protocol division: will fix.

File: src/api.py (hunk 1 above)
Comment (atyrode) [reply src/api.py]: Sorry, I don't think I understand what you mean, if it's null they get the standard pydantic return protocol as such:

```json
{
  "detail": [
    {
      "type": "missing",
      "loc": [
        "query",
        "value"
      ],
      "msg": "Field required",
      "input": null,
      "url": "https://errors.pydantic.dev/2.6/v/missing"
    }
  ]
}```

I assumed this was clear enough and didn't need re-implementation, but maybe I'm misunderstanding what you mean, could you provide a usage example that triggers a bug?

I got this from calling:

`/decode` and nothing else

File: src/api.py (hunk 2 above)
Comment (atyrode) [reply src/api.py]: was changed to a regex 
```python
rf"{DOMAIN_NAME}|{SHORT_URL}"
``` 
in https://github.com/atyrode/vite/pull/12/commits/499bebb26a581ec3e1a80184db1c5c5d681f0310

File: src/charset.py (hunk 4 above)
Comment (atyrode) [reply src/charset.py:66]: I'm okay with committing this suggestion but it joins my other comment below about readability over implicitness:
I agree it is slightly anti-pythonic but more welcoming and readable, and provides clear intent of code execution.

File: src/static/index.css (hunk 7 above)
Comment (atyrode) [reply src/static/index.css]: https://github.com/atyrode/vite/pull/12/commits/3c32e4953f7d3058d5631b405e926bd9a768a9a4

**Comment by atyrode** [comment]
There was a branch work in progress that was merged during the review so part of it might be outdated.
- #13 
into
- #14
into https://github.com/atyrode/vite/pull/12/commits/499bebb26a581ec3e1a80184db1c5c5d681f0310

File: src/api.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (atyrode) [reply src/api.py]: should be fully resolved by https://github.com/atyrode/vite/commit/d1417c31c71ffbd9c4ea66437ad09803c8082b04
and added instructions how to provide a .env in the readme with example values

File: src/database.py (hunk 6 above)
Comment (atyrode) [reply src/database.py:18]: now implemented after https://github.com/atyrode/vite/pull/12/commits/9821d35af01dc0f63374c413c59476fe5c12cf9c, thanks for mentionning SQLAlchemy, I had no idea this way a thing and I won't be looking back, feels like a must now 😄

**Comment by atyrode** [comment]
This should be the final state of the release candidate as the deadline for the client's need for this project is by midnight, so I consider it stable enough (passes test, all endpoints behaves as expected, everything is documented, and the most important suggestions were implemented: ORM for database operations as well as secret management through python-dotenv).

The release candidate is thereby ready for review unless I find a last minute major bug that I can fix by midnight.

**Comment by atyrode** [comment]
This should be the final state of the release candidate as the deadline for the client's need for this project is by midnight, so I consider it stable enough (passes test, all endpoints behaves as expected, everything is documented, and the most important suggestions were implemented: ORM for database operations as well as secret management through python-dotenv).

The release candidate is thereby ready for review unless I find a last minute major bug that I can fix by midnight.

File: .github/workflows/python-test.yml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+      uses: actions/setup-python@v4
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Run pytest
+      run: |
+        pytest
```
Comment (pierre-gorde) [inline .github/workflows/python-test.yml:26]: 👍

File: README.md [hunk 11]
```diff
@@ -1,2 +1,79 @@
...
+VITE_HOST=localhost:8080
+```
+
+3. Using Python 3.9 or up, run:
+```shell
+pip install -r requirements.txt
+python3 start.py
+```
+
+4. Voilà! As a French person, would say: "c'est allé [vite, lol](http://vite.lol/)"
```
Comment (pierre-gorde) [inline README.md:28]: 😆

File: src/static/index.html [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <title>vite!</title>
+    <link rel="stylesheet" type="text/css" href="static/index.css">
+    <!-- <script src="static/api.js"></script> -->
```
Comment (pierre-gorde) [inline src/static/index.html:6]: this comment can be remove

File: src/static/particles.js [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+
+async function resultSlide(uid) {
+    let viteResult = document.getElementById('vite_result');
+
+    function resetTransform() {
+        if (!first_time) {
+            this.style.transform = "translateY(0)";
+        }
+    }
```
Comment (pierre-gorde) [inline src/static/particles.js:44]: Even if `resetTransform` is not used elsewhere, it's still better to declare it outside of the `resultSlide` 
 function
